server/plugin.go

```diff
@@ -18,11 +35,215 @@ type Plugin struct {
+func (p *Plugin) connectTeamsBotClient() error {
+	p.msteamsBotClientMutex.Lock()
+	if p.msteamsBotClient == nil {
+		p.msteamsBotClient = msteams.NewBot(
+			p.configuration.TenantId,
+			p.configuration.ClientId,
+			p.configuration.ClientSecret,
+			p.configuration.BotUsername,
+			p.configuration.BotPassword,
+		)
+	}
+	err := p.msteamsBotClient.Connect()
+	if err != nil {
+		p.API.LogError("Unable to connect to the bot client", "error", err)
+		return err
+	}
+	p.msteamsBotClientMutex.Unlock()
+	return nil
+}
+
+func (p *Plugin) start() {
+	err := p.connectTeamsAppClient()
+	if err != nil {
+		p.API.LogError("Unable to connect to the msteams", "error", err)
+		return
+	}
+	err = p.connectTeamsBotClient()
+	if err != nil {
+		p.API.LogError("Unable to connect to the msteams", "error", err)
+		return
+	}
+
+	channelsLinkedData, appErr := p.API.KVGet("channelsLinked")
+	if appErr != nil {
+		p.API.LogError("Error getting the channels linked", "error", appErr)
+		return
+	}
+	channelsLinked := map[string]ChannelLink{}
+	err = json.Unmarshal(channelsLinkedData, &channelsLinked)
+	if err != nil {
+		p.API.LogError("Error getting the channels linked", "error", err)
+		return
+	}
+
+	p.channelsLinked = channelsLinked
+	p.subscriptionsToLinks = map[string]ChannelLink{}
+	ctx, stop := context.WithCancel(context.Background())
+	p.stopSubscriptions = stop
+	err = p.msteamsAppClient.ClearSubscriptions()
+	if err != nil {
+		p.API.LogError("Unable to clear all subscriptions", "error", err)
+	}
+	for _, link := range channelsLinked {
+		go p.subscribeToChannel(ctx, link)
+	}
+}
+
+func (p *Plugin) stop() {
+	if p.stopSubscriptions != nil {
+		p.stopSubscriptions()
+	}
+}
+
+func (p *Plugin) restart() {
+	p.stop()
+	p.start()
+}
+
+func (p *Plugin) OnActivate() error {
+	p.stopSubscriptions = func() {}
+
+	bot, appErr := p.API.CreateBot(&model.Bot{
+		Username:    botUsername,
+		DisplayName: botDisplayName,
+		Description: "Created by the MS Teams Sync plugin.",
+	})
+	if appErr != nil {
+		bot, appErr := p.API.GetUserByUsername(botUsername)
+		if appErr != nil {
+			return appErr
+		}
+		p.userID = bot.Id
+	} else {
+		p.userID = bot.UserId
+	}
+
+	err := p.API.RegisterCommand(createMsteamsSyncCommand())
+	if err != nil {
+		return err
+	}
+
+	go p.start()
+	return nil
+}
+
+func (p *Plugin) MessageHasBeenPosted(c *plugin.Context, post *model.Post) {
+	if post.Props != nil {
+		if _, ok := post.Props["msteams_sync_"+p.userID].(bool); ok {
+			return
+		}
+	}
+
+	channel, _ := p.API.GetChannel(post.ChannelId)
+
+	link, ok := p.channelsLinked[channel.TeamId+":"+post.ChannelId]
+	if !ok {
+		return
+	}
+
+	user, _ := p.API.GetUser(post.UserId)
+
+	p.Send(link, user, post)
+}
+
+func (p *Plugin) OnDeactivate() error {
+	p.stop()
+	return nil
+}
+
+func (p *Plugin) Send(link ChannelLink, user *model.User, post *model.Post) (string, error) {
+	p.API.LogDebug("\n\n\n=> Receiving message", "post", post)
+
+	// TODO: Replace this with a template
```

Comment:
This TODO here is known, I don't want to do it right away to see how it goes in terms of users expectations.